### DISCUSSION TITLE
Deploy Middleware to Azure

### DISCUSSION THREAD

**Discussion by MauriRojas** [middlewarehq/middleware#621] [General]
Hi everyone,
Is it possible to deploy Middleware to Azure?

The architecture I detected so far is:
1. Postgres in Container App or Azure resource
2. Redis in Container App or Azure resource
3. Sync server in Container App
4. Analytics server in Container App
5. Frontend, probably in Web App

**Comment by jayantbh** [comment]
Should be absolutely possible.

As long as the target instance can run docker pull, has enough storage and memory, and can expose the webserver port to public in some way, you should be able to run it just fine.

If that's not the case do let us know. 🙂

**Reply by MauriRojas** [reply]
I'm trying to provision it at the moment.
So far I have:
1. PostgresSQL Azure resource
2. Redis running in a Container App
3. Trying to run the sync server in an Web App

If I run into any blockers I'll ask here.

Thank you

**Reply by jayantbh** [reply]
If I understand correctly, it seems like you're trying to run all parts of Middleware separately.

Is that a requirement?
If no, you could run everything in a single machine via the quick start instructions in the readme.

**Reply by MauriRojas** [reply]
Is not a requirement.
Good idea! I may try that.

**Reply by MauriRojas** [reply]
I have provisioned and deployed:
1. PostgresSQL Azure resource
2. Redis Azure resource
3. Sync server in Container App
4. Analytics server in Container App
5. Frontend running locally

I initialized the Postgres DB by building DB Mate locally and connecting it to the Azure resource.
I am able to add a GitHub PAT using the frontend, create a team and select a repo.
However, the DORA metrics are not generated, and the frontend is in-progress state indefinitely.

I found in the logs the following error in the sync server: "Error syncing org repos for provider github: Source contains parsing errors: '/app/mhq/config/config.ini". I generated the encryption keys using the [generate_config_ini.sh](https://github.com/middlewarehq/middleware/blob/main/setup_utils/generate_config_ini.sh) file and included it in the container images.

The ini file has this format after updated by the script:
<img width="439" alt="image" src="https://github.com/user-attachments/assets/7fd9cc35-f6a7-48a1-9e49-59d08a366317" />

_The frontend stays in this state indefinitely. Almost sure because is waiting for the sync server that is failing_
<img width="482" alt="image" src="https://github.com/user-attachments/assets/135573b4-b3ab-4753-b703-fb9cbf2321b6" />

**Reply by jayantbh** [reply]
Hey @MauriRojas, I'll ask someone from the team to look at it.
I expect it to be a bit tricky to really debug since the whole thing is spread over a bunch of different environments.

In the mean time, you can try the containerized setup with these instructions.

```sh
docker volume create middleware_postgres_data
docker volume create middleware_keys
docker run --name middleware \
           -p 3333:3333 \
           -p 9696:9696 \
           -p 9697:9697 \
           -v middleware_postgres_data:/var/lib/postgresql/data \
           -v middleware_keys:/app/keys \
           -d middlewareeng/middleware:latest
docker logs -f middleware
```

**Reply by MauriRojas** [reply]
Solved it. First I had to add `allow_no_value=True` to the [config parser instance creation](https://github.com/middlewarehq/middleware/blob/main/backend/analytics_server/mhq/utils/cryptography.py#L22):
`config = configparser.ConfigParser(allow_no_value=True)`

After that, the error changed to "Error in get_access_token - Not a valid PEM post boundary". To fix this, I removed the line breaks in the config.ini file.
The original file that the script created was:
[KEYS]
SECRET_PRIVATE_KEY=LS0tLS1CRUdJTiBQUklWQVRF
JDjdEDIJEeidjWIWPODMW

When it should be only one line so the whole value was read. Like this:
[KEYS]
SECRET_PRIVATE_KEY=LS0tLS1CRUdJTiBQUklWQVRFJDjdEDIJEeidjWIWPODMW

I don't know if the intention of the file was to be generated with line breaks or not, but that solved the issue.
